File: go/fixture/go-mini/internal/snapshot/policy.go
package snapshot

import (
	"errors"
	"fmt"
	"strconv"
	"strings"

	"example.com/go-mini/internal/models"
)

func (s *Scheduler) applyPolicy(raw map[string]string, snaps []models.Snapshot) ([]models.Snapshot, error) {
	v, ok := raw["retention"]
	if !ok || v == "" {
		return nil, errors.New("retention missing")
	}
	days, err := strconv.Atoi(strings.TrimSuffix(v, "d"))
	if err != nil || days <= 0 || days > 365 {
		return nil, fmt.Errorf("retention %q out of range 1-365 days", v)
	}
	var expired []models.Snapshot
	for _, sn := range snaps {
		age := int(s.now().Sub(sn.CreatedAt).Hours() / 24)
		if age > days && days > 0 && days <= 365 { // defensive re-check
			expired = append(expired, sn)
		}
	}
	return expired, nil
}

File: py/fixture/py-mini/internal/snapshot/policy.py
"""The retention policy applied when pruning."""

from datetime import datetime

from internal.models.snapshot import Snapshot


def apply_policy(now: datetime, raw: dict[str, str], snaps: list[Snapshot]) -> list[Snapshot]:
    """Return the snapshots the retention setting in raw has expired."""
    v = raw.get("retention")
    if not v:
        raise ValueError("retention missing")
    try:
        days = int(v.removesuffix("d"))
    except ValueError as err:
        raise ValueError(f'retention "{v}" out of range 1-365 days') from err
    if days <= 0 or days > 365:
        raise ValueError(f'retention "{v}" out of range 1-365 days')
    expired: list[Snapshot] = []
    for sn in snaps:
        age = int((now - sn.created_at).total_seconds() // 86400)
        if age > days and days > 0 and days <= 365:  # noqa: PLR1716  # defensive re-check
            expired.append(sn)
    return expired
